merge.py

```diff
@@ -1,98 +1,303 @@
-    print(rebase_success)
-    post_error('Rebase was not successful. Please rebase against main and try again.', pr)
-    return False
-  print("Rebased")
-
-  # Check out target branch to here
-  checkout_success = call(['git', 'checkout', '{}/{}'.format(TARGET_REMOTE, TARGET_BRANCH)], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not checkout_success == 0:
-    post_error('Error checking out target branch: master. Please try again.', pr)
-    return False
-  print("Checked out Apache master.")
-
-  # Merge
-  merge_success = call(['git', 'merge', '--no-ff', '-m', 'This closes #{}'.format(pr), 'finish-pr-{}'.format(pr)], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not merge_success == 0:
-    post_error('Merge was not successful against target branch: master. Please try again.', pr)
-    return False
-  print("Merged successfully.")
-
-  print("Running mvn clean verify.")
-  # mvn clean verify
-  mvn_success = call(['mvn', 'clean', 'verify'], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not mvn_success == 0:
-    post_error('mvn clean verify against HEAD + PR#{} failed. Not merging.'.format(pr), pr)
-    return False
-
-  # git push (COMMENTED UNTIL MERGEBOT HAS PERMISSIONS)
-  #push_success = call(['git', 'push', 'apache', 'HEAD:master'], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  #if not push_success == 0:
-  #  post_error('Git push failed. Please try again.', pr)
-  #  return False
-  return True
-
-def set_up():
-  if not call(['mkdir', '/usr/local/google/home/jasonkuster/tmp']) == 0:
-    return False
-  return True
-
-
-def clean_up():
-  if not call(['rm', '-rf', '/usr/local/google/home/jasonkuster/tmp']) == 0:
-    return False
-  return True
 
+
+import abc
+import logging
+import os
+import shlex
+from subprocess import check_call, check_output, STDOUT
+from threading import Thread
+
+
+APACHE_GIT = 'https://git-wip-us.apache.org/repos/asf/{repo}.git'
+GITHUB_REPO_URL = 'https://github.com/{org}/{repo}.git'
+TMP_DIR_FMT = '/tmp/{dir}'
+
+
+def create_merger(config, work_queue):
+    """create_merger creates a merger of the type specified in the config.
+
+    Args:
+        config: A dictionary containing repository configuration.
+        work_queue: A Queue.Queue where new work items will be queued.
+    Returns:
+        Merger of type specified in configuration
+    Raises:
+        AttributeError: if passed an unsupported SCM type.
+    """
+    if config['scm_type'] == 'github':
+        return GitMerger(config, work_queue)
+    raise AttributeError(
+        'Unsupported SCM type: {}.'.format(config['scm_type']))
+
+
+class Merger(Thread):
+    """Merger is the base class for all mergers.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, config, work_queue):
+        self.config = config
+        l = logging.getLogger('{name}_merge_logger'.format(
+            name=self.config['name']))
+        log_fmt = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
```

Comment:
Can you add some comments on the logging configuration? I see multiple loggers, with varying names and such which is cluttering this code.